samza-kafka/src/main/java/org/apache/samza/system/kafka/KafkaSystemAdmin.java

```diff
@@ -298,24 +298,78 @@ public BoxedUnit apply(Exception exception, ExponentialSleepStrategy.RetryLoop l
   @Override
   public Map<SystemStreamPartition, SystemStreamMetadata.SystemStreamPartitionMetadata> getSSPMetadata(
       Set<SystemStreamPartition> ssps) {
+    return getSSPMetadata(ssps,
+        new ExponentialSleepStrategy(DEFAULT_EXPONENTIAL_SLEEP_BACK_OFF_MULTIPLIER,
+            DEFAULT_EXPONENTIAL_SLEEP_INITIAL_DELAY_MS, DEFAULT_EXPONENTIAL_SLEEP_MAX_DELAY_MS));
+  }
+
+  /**
+   * Given a set of SystemStreamPartition, fetch metadata from Kafka for each
+   * of them, and return a map from ssp to SystemStreamPartitionMetadata for
+   * each of them. This method will return null for oldest and newest offsets
+   * if a given SystemStreamPartition is empty. This method will block and
+   * retry indefinitely until it gets a successful response from Kafka.
+   * @param ssps a set of strings of SSP
+   * @param retryBackoff retry backoff strategy
+   * @return a map from ssp to sspMetadata which has offsets
+   */
+  public Map<SystemStreamPartition, SystemStreamMetadata.SystemStreamPartitionMetadata> getSSPMetadata(
```

Comment:
Can this be package private? Also, add a unit tests as well.

Reply:
Modified it to a private one and added unite test to test retry.